eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxow.hs

```diff
@@ -63,20 +85,81 @@ conwayWitsVKeyNeeded = getConwayWitsVKeyNeeded
 
 -- ================================
 
-type instance EraRuleFailure "UTXOW" (ConwayEra c) = BabbageUtxowPredFailure (ConwayEra c)
+data ConwayUtxowPredFailure era
+  = UtxoFailure (PredicateFailure (EraRule "UTXO" era))
+  | InvalidWitnessesUTXOW
+      ![VKey 'Witness (EraCrypto era)]
+  | -- | witnesses which failed in verifiedWits function
+    MissingVKeyWitnessesUTXOW
+      -- | witnesses which were needed and not supplied
+      !(Set (KeyHash 'Witness (EraCrypto era)))
+  | -- | missing scripts
+    MissingScriptWitnessesUTXOW
+      !(Set (ScriptHash (EraCrypto era)))
+  | -- | failed scripts
+    ScriptWitnessNotValidatingUTXOW
+      !(Set (ScriptHash (EraCrypto era)))
+  | -- | hash of the full metadata
+    MissingTxBodyMetadataHash
+      !(AuxiliaryDataHash (EraCrypto era))
+  | -- | hash of the metadata included in the transaction body
+    MissingTxMetadata
+      !(AuxiliaryDataHash (EraCrypto era))
+  | ConflictingMetadataHash
+      -- | hash of the metadata included in the transaction body
+      !(AuxiliaryDataHash (EraCrypto era))
+      -- | expected hash of the full metadata
+      !(AuxiliaryDataHash (EraCrypto era))
+  | -- | Contains out of range values (string`s too long)
+    InvalidMetadata
+  | -- | extraneous scripts
+    ExtraneousScriptWitnessesUTXOW
+      !(Set (ScriptHash (EraCrypto era)))
+  | MissingRedeemers
+      ![(PlutusPurpose AsItem era, ScriptHash (EraCrypto era))]
+  | MissingRequiredDatums
+      -- | Set of missing data hashes
+      !(Set (DataHash (EraCrypto era)))
+      -- | Set of received data hashes
+      !(Set (DataHash (EraCrypto era)))
+  | NotAllowedSupplementalDatums
+      -- | Set of unallowed data hashes
+      !(Set (DataHash (EraCrypto era)))
+      -- | Set of acceptable supplemental data hashes
+      !(Set (DataHash (EraCrypto era)))
+  | PPViewHashesDontMatch
+      -- | The PPHash in the TxBody
+      !(StrictMaybe (ScriptIntegrityHash (EraCrypto era)))
+      -- | Computed from the current Protocol Parameters
+      !(StrictMaybe (ScriptIntegrityHash (EraCrypto era)))
+  | -- | Set of transaction inputs that are TwoPhase scripts, and should have a DataHash but don't
+    UnspendableUTxONoDatumHash
+      (Set (TxIn (EraCrypto era)))
+  | -- | List of redeemers not needed
+    ExtraRedeemers ![PlutusPurpose AsIx era]
+  | -- | Embed UTXO rule failures
+    MalformedScriptWitnesses
+      !(Set (ScriptHash (EraCrypto era)))
+  | -- | the set of malformed script witnesses
+    MalformedReferenceScripts
+      !(Set (ScriptHash (EraCrypto era)))
+  deriving (Generic)
+
+type instance EraRuleFailure "UTXOW" (ConwayEra c) = ConwayUtxowPredFailure (ConwayEra c)
 
 type instance EraRuleEvent "UTXOW" (ConwayEra c) = AlonzoUtxowEvent (ConwayEra c)
 
-instance InjectRuleFailure "UTXOW" BabbageUtxowPredFailure (ConwayEra c)
+instance InjectRuleFailure "UTXOW" BabbageUtxowPredFailure (ConwayEra c) where
```

Comment:
With addition of a new predicate failure data type we need `InjectFailure` instances for all possible injections in their respective modules for that type for every rule that can inject it:

* `instance InjectRuleFailure "UTXOW" ConwayUtxowPredFailure (ConwayEra c)`
* `instance InjectRuleFailure "LEDGER" ConwayUtxowPredFailure (ConwayEra c)`
* `instance InjectRuleFailure "LEDGERS" ConwayUtxowPredFailure (ConwayEra c)`
* `instance InjectRuleFailure "BBODY" ConwayUtxowPredFailure (ConwayEra c)`